File: sort/quick_sort.py
def partition(arr, low, high):

    pivot = arr[high]

    left = low
    right = high-1
    while left <= right:
        if arr[left] <= pivot:
            left += 1
        elif arr[right] >= pivot:
            right -= 1
        # if not fit left and right criteria, swap them
        else:
            arr[left], arr[right] = arr[right], arr[left]
            left += 1
            right -= 1

    arr[left], arr[high] = arr[high], arr[left]
    return left


def quick_sort(arr, low, high):
    if low < high:
        partition_index = partition(arr, low, high)
        quick_sort(arr, low, partition_index - 1)
        quick_sort(arr, partition_index + 1, high)


def main():
    list = [99, 0, 5, 20, 123, 0, -1, 72, 21, 22, 13, 8, 7, 67, 29, 1, 2, 4]
    quick_sort(list, 0, len(list) - 1)
    print(list)

    list = [3, 9, 2, 1]
    quick_sort(list, 0, len(list) - 1)
    print(list)


main()

File: sort/quick_sort.js
const partition = (arr, low, high) => {
  let pivot = arr[high];
  let left = low;
  let right = high;

  while (left <= right) {
    if (arr[left] <= pivot) {
      left++;
    } else if (arr[right] >= pivot) {
      right--;
    } else {
      swap(arr, left, right);
      left++;
      right--;
    }
  }

  swap(arr, left, high);
  return left;
};

function quickSort(arr, L, R) {
  if (L < R) {
    const pivotIndex = partition(arr, L, R);
    quickSort(arr, L, pivotIndex - 1);
    quickSort(arr, pivotIndex + 1, R);
  }
}

function swap(arr, index1, index2) {
  // es6 Destructuring
  [arr[index1], arr[index2]] = [arr[index2], arr[index1]];
}

let data = [8, 9, 2, 5, 1];
quickSort(data, 0, data.length - 1);
console.log(data);
